src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/InsertTableOperation.java

```diff
@@ -82,14 +91,43 @@ private Uni<TableInsertAttempt> insertRow(
 
   private SimpleStatement buildInsertStatement(QueryExecutor queryExecutor, WriteableTableRow row) {
 
-    Map<CqlIdentifier, Term> colValues =
-        row.allColumnValues().entrySet().stream()
-            .collect(Collectors.toMap(Map.Entry::getKey, e -> literal(e.getValue())));
+    Preconditions.checkArgument(
+        !row.allColumnValues().isEmpty(), "Row must have at least one column to insert");
 
-    return insertInto(
-            commandContext.schemaObject().name.keyspace(),
-            commandContext.schemaObject().name.table())
-        .valuesByIds(colValues)
-        .build();
+    InsertInto insertInto =
+        insertInto(
+            commandContext.schemaObject().tableMetadata.getKeyspace(),
+            commandContext.schemaObject().tableMetadata.getName());
+
+    List<Object> positionalValues = new ArrayList<>(row.allColumnValues().size());
+    RegularInsert ongoingInsert = null;
+
+    for (Map.Entry<CqlIdentifier, Object> entry : row.allColumnValues().entrySet()) {
+      try {
+        var codec =
+            JSONCodecRegistry.codecToCQL(
+                commandContext.schemaObject().tableMetadata, entry.getKey(), entry.getValue());
+        positionalValues.add(codec.toCQL(entry.getValue()));
+      } catch (UnknownColumnException e) {
+        // TODO AARON - Handle error
+        throw new RuntimeException(e);
+      } catch (MissingJSONCodecException e) {
+        // TODO AARON - Handle error
+        throw new RuntimeException(e);
+      } catch (ToCQLCodecException e) {
+        // TODO AARON - Handle error
+        throw new RuntimeException(e);
+      }
+
+      // need to switch from the InertInto interface to the RegularInsert to get to the
+      // asCQL() later.
+      ongoingInsert =
+          ongoingInsert == null
+              ? insertInto.value(entry.getKey(), bindMarker())
+              : ongoingInsert.value(entry.getKey(), bindMarker());
+    }
+
+    assert ongoingInsert != null;
```

Comment:
I don't think we use (or should use Java `assert` keyword) -- `Objects.requireNonNull()` better.